include/wil/registry_basic.h

```diff
@@ -0,0 +1,1602 @@
+        {
+            return ::wil::reg::set_value_nothrow(key, nullptr, value_name, data);
+        }
+
+        inline HRESULT set_value_expanded_string_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, _In_ PCWSTR data) WI_NOEXCEPT
+        {
+            const reg_view_details::reg_view_nothrow regview{ key };
+            return regview.set_value_with_type(subkey, value_name, data, REG_EXPAND_SZ);
+        }
+
+        inline HRESULT set_value_expanded_string_nothrow(HKEY key, _In_opt_ PCWSTR value_name, _In_ PCWSTR data) WI_NOEXCEPT
+        {
+            return ::wil::reg::set_value_expanded_string_nothrow(key, nullptr, value_name, data);
+        }
+
+#if defined(_VECTOR_) && defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+        inline HRESULT set_value_multistring_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_ PCWSTR value_name, const ::std::vector<::std::wstring>& data) WI_NOEXCEPT
+        {
+            const reg_view_details::reg_view_nothrow regview{ key };
+            return regview.set_value_multistring(subkey, value_name, data);
+        }
+
+        inline HRESULT set_value_multistring_nothrow(HKEY key, _In_ PCWSTR value_name, const ::std::vector<::std::wstring>& data) WI_NOEXCEPT
+        {
+            return ::wil::reg::set_value_multistring_nothrow(key, nullptr, value_name, data);
+        }
+#endif // #if defined(_VECTOR_) && defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+
+        //
+        // get_value* (throwing) functions
+        //
+#if defined(WIL_ENABLE_EXCEPTIONS)
+        template <typename T>
+        T get_value(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name)
+        {
+            const reg_view_details::reg_view regview{ key };
+            T return_value{};
+            regview.get_value<T>(subkey, value_name, return_value);
+            return return_value;
+        }
+
+        template <typename T>
+        T get_value(HKEY key, _In_opt_ PCWSTR value_name)
+        {
+            return ::wil::reg::get_value<T>(key, nullptr, value_name);
+        }
+
+        inline DWORD get_value_dword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name)
+        {
+            return wil::reg::get_value<DWORD>(key, subkey, value_name);
+        }
+
+        inline DWORD get_value_dword(HKEY key, _In_opt_ PCWSTR value_name)
+        {
+            return wil::reg::get_value<DWORD>(key, nullptr, value_name);
+        }
+
+        inline DWORD64 get_value_qword(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name)
+        {
+            return wil::reg::get_value<DWORD64>(key, subkey, value_name);
+        }
+
+        inline DWORD64 get_value_qword(HKEY key, _In_opt_ PCWSTR value_name)
+        {
+            return wil::reg::get_value<DWORD64>(key, nullptr, value_name);
+        }
+
+#if defined(_STRING_)
+        inline ::std::wstring get_value_wstring(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name)

```

Comment:
I may have left this comment before, but we should avoid encoding type information in the function name.

We should have `get_value_string` that by default returns `std::wstring`, and then maybe templated versions that take `bstr` or `cotaskmem_string`.

Reply:
Especially since we want the variants (expanded, multi, etc) to work for all string types.

Reply:
I hear you. I have also heard from multiple sources to be more explicit. I'm happy with the current balance.